cfme/tests/networks/test_sdn_inventory_collection.py

```diff
@@ -1,30 +1,76 @@
-from cfme.cloud.provider.openstack import OpenStackProvider
-from cfme.networks.cloud_network import CloudNetwork
-from cfme.utils.appliance.implementations.ui import navigate_to
-
+from cfme.cloud.provider.gce import GCEProvider
+from cfme.utils.blockers import BZ
 
 pytestmark = [
-    pytest.mark.usefixtures('setup_provider'),
-    pytest.mark.provider([EC2Provider, AzureProvider, OpenStackProvider],
-                         scope='module'),
+    pytest.mark.provider([EC2Provider, AzureProvider, GCEProvider]),
+    pytest.mark.usefixtures('setup_provider')
 ]
 
 
-@pytest.mark.tier(1)
-@pytest.mark.uncollect
-def test_sdn_inventory_subnets(provider):
-    """ Test to compare mgmt system and ui list of networks
-    Note: Those providers don't have list_network() method implemented yet
-    """
-    network_names = provider.mgmt.list_network()
-    for network_name in network_names:
-        temp_network = CloudNetwork(name=network_name)
-        view = navigate_to(temp_network, 'Details')
-        net_name = view.entities.properties.get_text_of('Name')
-        assert net_name == network_name
-
-    provider.delete_if_exists(cancel=False)
-    provider.wait_for_delete()
+def test_sdn_api_inventory_networks(provider, appliance):
+    prov_networks = sorted(provider.mgmt.list_network())
+    appliance.server.browser.refresh()
+    cfme_networks = sorted([nt.name for nt in appliance.collections.cloud_networks.all()])
+
+    if provider.one_of(EC2Provider):
+        # Ec2 API returns only the Id of the networks, so the alternative is to count the networks
+        # instead of checking there names. Less accurate, but Better than nothing...
+        assert len(cfme_networks) == len(prov_networks), 'There is NOT the same amount of networks'
+        'in CFME than on EC2: {prov} {cfme}'.format(prov=prov_networks, cfme=cfme_networks)
+    else:
+        assert cfme_networks == prov_networks, 'Prov networks list: {networks} different from '
+        'cfme list: {cfme}'.format(networks=prov_networks, cfme=cfme_networks)
+
+
+@pytest.mark.provider([AzureProvider], override=True, scope='function')
+@pytest.mark.meta(blockers=[BZ(1550605, forced_streams=["5.9", "5.8"],
+                  unblock=lambda provider: not provider.one_of(AzureProvider))])
+@pytest.mark.uncollectif(lambda provider, appliance: provider.one_of(EC2Provider) and
+                         appliance.version < "5.9", reason='RFE for 5.9')
+def test_sdn_api_inventory_routers(provider, appliance):
+    prov_routers = sorted(provider.mgmt.list_router())
+    appliance.server.browser.refresh()
+    cfme_routers = sorted([rt.name for rt in appliance.collections.network_routers.all()])
+
+    assert cfme_routers == prov_routers, 'Prov routers list: {router} different from cfme list: '
+    '{cfme}'.format(router=prov_routers, cfme=cfme_routers)
+
+
+def test_sdn_api_inventory_subnets(provider, appliance):
+    prov_subnets = []
+    appliance.server.browser.refresh()
```

Comment:
QUESTION: what is it for ?

Reply:
Because when running ``appliance.collections.network_subnets.all()``, It returns the list from UI for the first provider. Then for the next provider, it is not refreshed ( probably because the is_displayed  returns True) and then the network_subnets of the previous provider are still displayed and the whole test fails.
If you run --use-provider=azure it works if you run ==use-provider=complete it does not. 
Thats why the page refresh.

Reply:
Please add a short comment in-line noting why this is done. Also look at the `resetter()` method for the navigation destination, and consider adding a refresh call to it instead, so that even if is_displayed is true and it doesn't navigate, it will refresh.

Reply:
@t-fischer, shouldn't this refresh be called in all() ?